File: 백준/Silver/11403. 경로 찾기/경로 찾기.py
import sys
input = sys.stdin.readline

n = int(input())
graph = [list(map(int, input().split())) for _ in range(n)]

edges = {}
for i in range(n):
    edges[i] = []

for i in range(n):
    for j in range(n):
        if graph[i][j]:
            edges[i].append(j)

matrix = [[0] * n for _ in range(n)]

def dfs(v):
    for w in edges[v]:
        if not visited[w]:
            visited[w] = True
            dfs(w)

for i in range(n):
    visited = [False] * n
    dfs(i)
    for j in range(n):
        if visited[j]:
            matrix[i][j] = 1

for i in range(n):
    print(*matrix[i])

File: 백준/Silver/11403. 경로 찾기/경로 찾기.java
import java.util.*;
import java.io.*;


public class Main {
    static int n;
    static Map<Integer, List<Integer>> edges;
    static boolean[] visited;

    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        n = Integer.parseInt(br.readLine());
        StringTokenizer st;

        edges = new HashMap<>();
        for (int i = 0; i < n; i++) {
            edges.put(i, new ArrayList<>());
        }

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                if (st.nextToken().equals("1")) {
                    edges.get(i).add(j);
                }
            }
        }

        int[][] matrix = new int[n][n];
        for (int i = 0; i < n; i++) {
            visited = new boolean[n];
            dfs(i);
            for (int j = 0; j < n; j++) {
                if (visited[j]) {
                    matrix[i][j] = 1;
                }
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                bw.write(matrix[i][j] + " ");
            }
            bw.write('\n');
        }

        bw.flush();
        bw.close();
        br.close();
    }

    static void dfs(int v) {
        for (int w : edges.get(v)) {
            if (!visited[w]) {
                visited[w] = true;
                dfs(w);
            }
        }
    }
}